sql/grant.go

```diff
@@ -64,43 +45,21 @@ func (p *planner) Grant(n *parser.Grant) (planNode, error) {
 		return nil, err
 	}
 
-	// Let's just build a map of users->permissions and turn it back into two lists.
-	readers := map[string]struct{}{}
-	writers := map[string]struct{}{}
-	for _, readUser := range dbDesc.Read {
-		readers[readUser] = struct{}{}
-	}
-	for _, writeUser := range dbDesc.Write {
-		writers[writeUser] = struct{}{}
+	permissions, err := PermissionsFromDescriptor(dbDesc)
+	if err != nil {
+		return nil, err
 	}
 
-	// Now set the grantees.
-	for _, grantee := range n.Grantees {
-		if grantee == security.RootUser {
-			return nil, fmt.Errorf("grant: cannot modify permissions for %s", security.RootUser)
-		}
-		if isRead {
-			readers[grantee] = struct{}{}
-		}
-		if isWrite {
-			writers[grantee] = struct{}{}
-		}
+	if err := permissions.Grant(n); err != nil {
+		return nil, err
 	}
 
-	// Clear the permissions and re-build from the map.
+	// Copy the object. Permissions.ToDescriptor will overwrite the permission fields.
 	newDesc := *dbDesc
-	newDesc.Read, newDesc.Write = []string{}, []string{}
-	for user := range readers {
-		newDesc.Read = append(newDesc.Read, user)
-	}
-
-	for user := range writers {
-		newDesc.Write = append(newDesc.Write, user)
+	if err := permissions.FillDescriptor(&newDesc); err != nil {
+		return nil, err
 	}
 
-	sort.Strings(newDesc.Read)
-	sort.Strings(newDesc.Write)
-
 	// Now update the descriptor.
 	descKey := keys.MakeDescMetadataKey(newDesc.ID)
 	if err := p.db.CPut(descKey, &newDesc, dbDesc); err != nil {
```

Comment:
A conditional put seems sub-optimal. Shouldn't this be a transaction instead, so it's properly retried and doesn't just fail if there's concurrent access?